whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,414 @@
+This works well, but it will become ugly when we add more alternatives to the computation  f.e. a `isCharly :: String -> Maybe String`.
+
+It would be nice if we could have an operator to write it in a nice way like this:
+
+```haskell
+isCharly :: String -> Maybe String
+isCharly a = if a == "charly" then Just a else Nothing
+
+isAliceOrBobOrCharly :: String -> Maybe String
+isAliceOrBobOrCharly a = isAlice a <|> isBob a <|> isCharly a
+-- or
+isAliceOrBobOrCharly a = isAliceOrBob a <|> isCharly a
+```
+
+
+## The Alternative typeclass
+
+Actually, the code above is valid Haskell. All you need to do to make use of the `<|>` operator is to import it from `Control.Aplicative`:
+
+```none
+λ> :info (<|>)
+type Alternative :: (* -> *) -> Constraint
+class Applicative f => Alternative f where
+  ...
+  (<|>) :: f a -> f a -> f a
+  ...
+  	-- Defined in ‘GHC.Base’
+infixl 3 <|>
+```
+
+It's a method of the `Alternative` typeclass. Let's look at what information GHCI can provide us about this typeclass:
+
+```haskell
+type Alternative :: (* -> *) -> Constraint
+class Applicative f => Alternative f where
+  empty :: f a
+  (<|>) :: f a -> f a -> f a
+  some :: f a -> f [a]
+  many :: f a -> f [a]
+  {-# MINIMAL empty, (<|>) #-}
+  	-- Defined in ‘GHC.Base’
+instance Alternative ZipList -- Defined in ‘Control.Applicative’
+instance Alternative [] -- Defined in ‘GHC.Base’
+instance Alternative Maybe -- Defined in ‘GHC.Base’
+instance Alternative IO -- Defined in ‘GHC.Base’
+```
+
+This already gives us a lot of useful facts:
+
+1. A type implementing `Alternative` must be of kind `* -> *`.
+2. It must also implement `Applicative` (and, with that, also `Functor`).
+3. To implement it, you must implement `<|>` and `empty`.
+4. In addition to those methods, it gives you `some` and `many`, which we'll discuss further below.
+5. It's already defined for the `Maybe` type that we used in the example above.
+
+Lets take a look on the implementation of `Alternative Maybe`:
+
+```haskell
+instance Alternative Maybe where
+    empty = Nothing
+    Nothing <|> r = r
+    l       <|> _ = l
+```
+
+If you take a look at the definitions you can see that `empty :: f a` is
+the identity of `<|>` (very much like `mempty :: a` is the identity of
+`mappend :: a -> a -> a` in case of `Monoid`.) `Alternative` is also
+said to be _a monoid on applicative functors_.
+
+The truth table of `Alternative Maybe` looks like this:
+
+|             | **Nothing** | **Just r** |
+|-------------|-------------|------------|
+| **Nothing** | Nothing     | Just r     |
+| **Just l**  | Just l      | Just l     |
+
+
+Another interesting `Alternative` instance is `IO`.
+Here, the first alternative from left to right that succeeds wins, f.e.
+
+```haskell
+test :: IO String
+test = fail "foo" <|> pure "bar" <|> fail "baz"
+```
+
+When executed the fist computation `fail "foo"` obviously fails, so the second one `pure "bar"` succeeds, which is why the third `fail "baz"` isn't executed at all.
```

Comment:
s/fist/first/

Reply:
although I'd like the idea of reasoning about what a "fist computation` might be, I fixed it ;)